### PR TITLE
Constant-fold expressions that are default nulls

### PR DESCRIPTION
**Pull request by kagamiori**
Summary: When an expression has default null behavior and has at least one input being a constant NULL, we constant-fold this expression by replacing it with a constant NULL.

This PR is for https://github.com/facebookincubator/velox/issues/3072.

Differential Revision: D37809170

### CODE REVIEW

File: velox/expression/ConstantExpr.h [hunk 1]
```diff
@@ -57,6 +57,10 @@ class ConstantExpr : public SpecialForm {
     return sharedSubexprValues_;
   }
 
+  bool isNull() const {
```
Comment (mbasmanova) [inline velox/expression/ConstantExpr.h:60]: Would this work for complex types, i.e. when expression is created using ConstantExpr(VectorPtr value) constructor?

File: velox/expression/ExprCompiler.cpp [hunk 2]
```diff
@@ -275,12 +275,24 @@ ExprPtr tryFoldIfConstant(const ExprPtr& expr, Scope* scope) {
...
+      bool defaultNull = false;
       for (auto& input : expr->inputs()) {
-        if (!dynamic_cast<ConstantExpr*>(input.get())) {
-          return expr;
+        auto* constantInput = dynamic_cast<ConstantExpr*>(input.get());
+        if (!constantInput) {
+          allLiteral = false;
+        } else if (expr->propagatesNulls() && constantInput->isNull()) {
+          defaultNull = true;
+          break;
```
Comment (mbasmanova) [inline velox/expression/ExprCompiler.cpp:288]: Can we just return null constant expression here?

File: velox/expression/tests/ExprTest.cpp [hunk 3]
```diff
@@ -1072,6 +1101,23 @@ TEST_F(ExprTest, constantFolding) {
     auto typedExpr = parseExpression("codepoint('abcdef')");
     EXPECT_NO_THROW(exec::ExprSet exprSet({typedExpr}, execCtx_.get(), true));
   }
+
+  // Test constant-folding when the expression has default null behavior and at
+  // least one input is null.
+  VELOX_REGISTER_VECTOR_FUNCTION(udf_default_null, "default_null");
+  typedExpr = parseExpression("sign(default_null(null, 0.1))");
```
Comment (mbasmanova) [inline velox/expression/tests/ExprTest.cpp:1108]: I wonder if this expression would be constant folded even without the changes in this PR. I think we need to use an expression where not all inputs are constants, e.g. sign(floor(null) + c0)

**Review by mbasmanova — COMMENTED** [review]
@kagamiori Wei, thank you for working on more powerful constant folding. Some comments.

File: velox/expression/ConstantExpr.h (hunk 1 above)
Comment (kagamiori) [reply velox/expression/ConstantExpr.h:60]: Thank you for catching this! I set a new flag isNull_ in both constructors in the new version.

File: velox/expression/tests/ExprTest.cpp (hunk 3 above)
Comment (kagamiori) [reply velox/expression/tests/ExprTest.cpp:1108]: You're right. I fixed this expression in the new version.

**Pull request closed without merging**